react-1/src/components/User.js

```diff
@@ -0,0 +1,119 @@
+        float: 'left'
+    },
+    Arrow: {float: 'right'}
+});
+
+const calc = (x, y) => [-(y - window.innerHeight / 2) / 20, (x - window.innerWidth / 2) / 20, 1.1];
+const trans = (x, y, s) => `perspective(200px) scale(${s})`;
+
+const User = React.forwardRef(
+    ({user, isDragging, connectDragSource, connectDropTarget}, ref) => {
+        const classes = useStyles({user, isDragging, connectDragSource, connectDropTarget});
+        const elementRef = useRef(null);
+        const opacity = Number(!isDragging);
+	    const [props, set] = useSpring(() => ({ xys: [0, 0, 1], config: { mass: 5, tension: 250, friction: 20 } }));
+	    connectDragSource(elementRef);
+        connectDropTarget(elementRef);
+        useImperativeHandle(ref, () => ({
+            getNode: () => elementRef.current,
+        }));
+        return (
+	        <animated.div className={classes.Container}
+		        onMouseDown={({ clientX: x, clientY: y }) => set({ xys: calc(x, y) })}
+		        onMouseUp={() => set({ xys: [0, 0, 1] })}
+		        onMouseOver={({ clientX: x, clientY: y }) => set({ xys: calc(x, y) })}
+		        onMouseLeave={() => set({ xys: [0, 0, 1] })}
+		        onDragOver={({ clientX: x, clientY: y }) => set({ xys: calc(x, y) })}
+		        onDragLeave={() => set({ xys: [0, 0, 1] })}
+		        style={{ transform: props.xys.interpolate(trans) }}>
+            <li className={classes.User} style={{opacity}} ref={elementRef}>
+                <div className={classes.Infos}>
+                    <span className={classes.UserName}>{user.name}</span>
+                    <br/>
+                    <a href={`mailto:${user.email}`} className={classes.UserEmail}>{user.email}</a>
+                </div>
+                <img src="./arrow.svg" width="42" alt="➡️" className={classes.Arrow}/>
+            </li>
+	        </animated.div>
+        )
+    },
+);
+
+export default DropTarget(
+    ItemTypes.USER,
+    {
+        hover(props, monitor, component) {
+            if (!component) {
+                return null
+            }
+            const node = component.getNode();
+            if (!node) {
+                return null
+            }
+            const dragIndex = monitor.getItem().index;
+            const hoverIndex = props.index;
+            if (dragIndex === hoverIndex) {
+                return
+            }
+            const hoverBoundingRect = node.getBoundingClientRect();
+            const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+            const clientOffset = monitor.getClientOffset();
+            const hoverClientY = clientOffset.y - hoverBoundingRect.top;
+            if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+                return
+            }
+            if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
+                return
+            }
+            props.moveUser(dragIndex, hoverIndex);
+            monitor.getItem().index = hoverIndex;
+        },
```

Comment:
Some comments would help to understand what we are doing here.

Reply:
This procedure is launched at the beginning of a drop event to determine if the item dragged is in a droppable zone and is not over his own place, to determine which item he is going to replace and change their index.